Added `quantile` function

supreet11agrawal opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Added `quantile` function in `stats`. It takes the distribution and `p` as parameter and outputs the quantile.

#### Other comments
I haven't added median in this PR. It will just be a special case of `quantile` with `p = S(1)/2`. Tell me if I should add it as well

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * Added `quantile` function
<!-- END RELEASE NOTES -->

supreet11agrawal committed:
Added `quantile` function

supreet11agrawal committed:
Python2.7 tests

supreet11agrawal committed:
Imported `erfinv`

supreet11agrawal committed:
Added quantile support for discrete distributions

supreet11agrawal commented on sympy/stats/rv_interface.py:
```diff
@@ -208,6 +209,59 @@ def skewness(X, condition=None, **kwargs):
+    _value_check(p > 0, "The order p must be positive.")
+    _value_check(p < 1, "The order p must be less than 1.")
+
+    if pspace(X).is_Continuous:
```
Maybe here, I can write `if  pspace(X).is_Conitnuous or pspace(X).is_Discrete:`
It will cover discrete random variables as well in that case

supreet11agrawal replied:
No, it will be wrong. I'll think of something else

supreet11agrawal replied:
I have added support for discrete cases as well

supreet11agrawal commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -351,6 +356,7 @@ def test_exponential():
     assert P(X > 0) == S(1)
     assert P(X > 1) == exp(-rate)
     assert P(X > 10) == exp(-10*rate)
+    assert quantile(X, p) == Intersection(Interval(0, oo), FiniteSet(-log(1 - p)/rate))
```
`quantile(X)(x)` is a not a function of `x`. Maybe it will make more sense to define `quartiles` on the basis of these. I am not sure though
Or maybe we can make `quantile(X)(p)`. It is feasible, I think

supreet11agrawal replied:
I looked into it and actually `quantile(X)(x)` or `quantile(X)(p)` won't make much sense.

Upabjojr replied:
The quantile function accepts one parameter.

supreet11agrawal commented on sympy/stats/rv_interface.py:
```diff
@@ -208,6 +209,59 @@ def skewness(X, condition=None, **kwargs):
+    \6   for p < 1
+    """
+    _value_check(p > 0, "The order p must be positive.")
+    _value_check(p < 1, "The order p must be less than 1.")
```
Yes, reference: https://en.wikipedia.org/wiki/Quantile

Upabjojr replied:
It makes sense to call `quantile(X)(0)` or `quantile(X)(1)`.

supreet11agrawal replied:
Actually, p=1 has no meaning in reference in case of quantile (except finite cases). Consider exponential distribution. The quantile is denoted as `(-log(1-p))/rate`. For p = 1, numerator has log(0). In this case, however, p=0 seems valid.
This result comes directly from the distribution function. For exponential distribution, the domain is [0, oo). That is why p=0 gives results in this case. For most of the distributions, the domain is (-oo, +oo). In such cases quantile doen't exist for both p=0 and p=1.
For discrete cases, p=0 always exists but it is not accepted worldwide.

Upabjojr replied:
It will return oo.

supreet11agrawal replied:
Yes, that would be right for limiting case. But explicitly p=1 has no meaning. Perhaps, one can use `limit` for such cases

supreet11agrawal replied:
Maybe I should add explaination for such limiting cases in docs and illustrate the same by adding examples to the docs.

Upabjojr replied:
Why all this hussle? We support returning infinity, look at the tangent function and pass pi/2. Limits are a separate concept.

supreet11agrawal replied:
Ok, I'll do the changes

supreet11agrawal committed:
Added separate case for discrete distributions
However, this method is limited by Sympy's capability of solving inequalities

supreet11agrawal commented:
Should I add `median` function as well?

supreet11agrawal committed:
Added boundary conditions

supreet11agrawal committed:
Changed error message strings for boundary conditions.

supreet11agrawal committed:
In quantile, added cases for p=1

Upabjojr commented:
Hi, the `quantile` function is still accepting two parameters. It should accept only one.

supreet11agrawal commented:
I am not able to understand. It should take the distribution for which quantile is being asked and `p` as the order of quantile. Shouldn't it?

supreet11agrawal commented:
All the properties related to stats module such as `moment`, `variance`, `cmoment`, `covariance` work in the similar manner.

Upabjojr commented:
```python
In [8]: X = Normal("X", 0, 1)                                                                                                                                                                               

In [9]: density(X)(x)                                                                                                                                                                                       
Out[9]: 
      2 
    -x  
    ────
     2  
√2⋅ℯ    
────────
  2⋅√π  

In [10]: cdf(X)(x)                                                                                                                                                                                          
Out[10]: 
   ⎛√2⋅x⎞    
erf⎜────⎟    
   ⎝ 2  ⎠   1
───────── + ─
    2       2
```

`density` and `cdf` are correctly implemented as functions of one parameter. So should the `quantile`.

Upabjojr commented:
> All the properties related to stats module such as `moment`, `variance`, `cmoment`, `covariance` work in the similar manner.

It's because they are not functions.

supreet11agrawal commented:
Oh ok! Now I get it. I will change `quantile` function accordingly.

supreet11agrawal committed:
Added quantile for continuous cases.
Test cases and docs still needed to be corrected.
Further work - Adding quantile for finite and discrete cases

supreet11agrawal committed:
Completed quantile implementation for all distributions
Discrete distributions need help though

supreet11agrawal committed:
Merge branch 'master' into impl_quantile

supreet11agrawal committed:
Update __init__.py

supreet11agrawal commented:
This error is due to some other PR's change in `cdf` of Cauchy distribution. I will fix this

supreet11agrawal committed:
Update crv_types.py

supreet11agrawal commented:
@Upabjojr Are the changes Ok?

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -288,6 +288,32 @@ def expectation(self, expr, var, evaluate=True, **kwargs):
+
+    def quantile(self, x, **kwargs):
+        """ Cumulative density function """
+        if len(kwargs) == 0:
```
why can't `_quantile` have `kwargs`?

Upabjojr commented:
I guess this PR is good to merge. Thanks for your contribution.
